### PR TITLE
fix(cli): write prove outputs alongside the proof file

### LINKED ISSUE #3737: bug: prove --proof writes the outputs file where verify will not look for it
miden-vm 0.31.0 on `next`, the code is in `miden-vm/src/cli/prove.rs`.

`prove` writes two files: the proof itself, and an outputs file that only exists so `verify` can read it back. If you pass `--proof`, those two end up in different places. The proof goes where you asked, but the outputs file still gets written next to the program (`prove.rs:192`). `verify` works out where the outputs are by taking the proof path and swapping the extension (`verify.rs:109-116`), so it looks in a directory the file was never written to.

Proving to a path of your choosing and then verifying that proof doesn't work, and the error points at a file you never made:

```
$ miden-vm prove prog.masm --proof out/custom.proof
Program proved in 1298 ms

$ miden-vm verify -p out/custom.proof -x 8992a530f0d7...
Error:   x Failed to open input file out/custom.inputs
```

You can see this yourself by proving any program with `--proof out/custom.proof` and then listing the files. The proof is in `out/`, the outputs are sitting next to the program. If you point verify at them by hand with `-o prog.outputs` it verifies at 96 bits, so there's nothing wrong with the proof, just where the file was put. Without `--proof` both paths come out the same and you never notice.

The inputs file is a different case, since you write that one yourself rather than the CLI generating it, so having to pass `-i` to verify seems reasonable and I haven't touched it.

I'd open a PR that takes the default outputs path from the proof path rather than the program path, with tests, if someone assigns this to me. Without `--proof` the proof path already falls back to the program path, so nothing changes for anyone not using the flag.

### PR DESCRIPTION
**Pull request by rayasa07**
## Describe your changes

`prove` writes an outputs file so that `verify` can read the stack outputs back. `verify` locates it by taking the proof path and swapping the extension (`verify.rs:109-116`), but `prove` derived it from the *program* path instead. The two only coincide when the proof happens to sit in the program's directory under the program's basename, so any other `--proof` path left the outputs somewhere `verify` would not look.

This derives the default outputs path from the resolved proof path instead. `ProofFile` owns one `resolve_path` that both the proof write and the output derivation use, so the two cannot drift apart again. Without `--proof` the proof already falls back to the program path, so the two expressions reduce to the same thing and existing behaviour is unchanged.

Deriving the outputs from the proof makes one degenerate case reachable for any `.outputs` proof path: it resolves to itself, so writing the outputs would replace the proof. An explicit `--output` that repeats the proof path fails the same way and is pre-existing on `next`. `prove` now refuses to write the outputs when the two paths name the same file, after the proof has been written, so the proof is what survives the collision.

The check is `canonicalize` on both final paths. Running it after `ProofFile::write` is what makes that possible: the proof exists by then, so the filesystem can answer the question directly instead of the CLI predicting the answer. Symlinks, `.` and `..` components, case-insensitive names and Unicode normalization all fold away in the one call, and the guard fires only where the filesystem genuinely aliases the two paths rather than wherever a comparison guesses it might. Hard links stay invisible, because they share no resolved path.

Closes #3737

## Rationale

The outputs file is an internal handoff between `prove` and `verify`, not something a user is expected to manage, so the two commands have to agree on where it lives. They did not: `prove` anchored it to the program and `verify` anchored it to the proof. Whenever those anchors resolved differently the round trip broke, and the error named a path the user had never written to.

Anchoring the outputs to the proof makes the two sides agree by construction, and it is the anchor that makes sense on its own terms, since the outputs file is only ever read back together with the proof it belongs to.

The collision guard is included because this change is what makes the collision reachable for any `.outputs` proof path rather than just one. Replacing a proof the user just waited for is worse than the bug being fixed.

Earlier revisions of this PR ran the guard before any proving work, which meant predicting which distinct spellings the filesystem would treat as one file. Each round of review found another spelling that prediction missed. Asking the filesystem after the proof exists replaces all of that with one `canonicalize` per side and removes a 39-line resolver. The cost is that a collision is now reported after proving rather than before, and the benefit is that the proof is on disk when it is reported: the user re-runs with a different `--output` to get the outputs, and does not lose the proof either way.

One behaviour follows from delegating to the filesystem and is worth stating. A pair such as `custom.OUTPUTS` and `custom.outputs` is one file on a case-insensitive filesystem and two files on ext4, so it is now rejected on the first and accepted on the second. That is the same rule applied consistently, not a host-dependent guess.

The inputs file is a separate case and is deliberately untouched. The user writes that one by hand rather than the CLI generating it, so `prove` has nothing to relocate.

## Test plan

Unit tests cover the path computation and the resolver; CLI integration tests cover the call site, so the behaviour this PR fixes is pinned end to end rather than only in the helper.

Unit tests (`prove.rs`): the default outputs path for no `--proof`, a custom `--proof` path, an extensionless proof path and an `.outputs` proof path; that the resolver sees through a symlinked proof path and through `..`; and that an outputs path which does not exist yet never matches.

Integration tests (`tests/integration/cli/cli_test.rs`, using the existing `assert_cmd` harness): the outputs land next to a custom proof and not next to the program; an `.outputs` proof path keeps the proof and refuses the outputs; the same for an explicit `--output` repeating the proof path, in relative and absolute form, through `.`/`..`, through a symlinked proof path, through a symlink the outputs path resolves to once the proof is written, and where `..` crosses a symlinked directory; an `.outputs` proof path is accepted when `--output` sends the outputs elsewhere. The case-only and Unicode-normalization pairs probe the filesystem under the test directory first and assert the branch that filesystem actually implements.

Verified by running, not by inspection:

- `cargo test --release -p miden-vm --features executable --bin miden-vm`: 14 passed, 0 failed.
- `cargo test --release -p miden-vm --features executable --test miden-cli`: 299 passed, 0 failed, 1 ignored.
- **Negative control, the guard:** making `resolve_to_same_file` return `false` while leaving everything else in place fails exactly the seven collision tests, each on `Unexpected success` with the proof replaced by a 201-byte outputs file. The case-only and Unicode tests keep passing, because on ext4 those two pairs are genuinely separate files and the guard is not what decides them.
- **Negative control, the fix:** restoring the pre-fix `self.program_file.with_extension("outputs")` fails `prove_writes_outputs_next_to_a_custom_proof_file` on "the outputs belong next to the proof, which is where `verify` looks for them", and fails the custom-path, extensionless and `.outputs` unit tests (`left: "dir/program.outputs"`, `right: "out/custom.outputs"`). The no-`--proof` unit test still passes, which is the invariant that existing behaviour is unchanged.
- **Manual end to end:** every alias from this review, re-run against the built binary. Each is rejected with exit 1 and a 39772-byte proof still on disk: `sub/../same.proof` against `same.proof`; `same.proof` against `$PWD/same.proof`; a dangling `custom.outputs -> custom.proof` with the default outputs, which also leaves the alias a symlink; `alias/../same.proof` with `alias -> real/sub`; `--output` repeating `--proof` exactly; and `--proof custom.outputs` with the default outputs. `prove program.masm --proof out/custom.proof` writes `out/custom.proof` and `out/custom.outputs` together with no stray `program.outputs`, and with no `--proof` the proof and outputs still land next to the program.
- `cargo +nightly fmt --all --check` and `cargo +1.98.1 clippy -p miden-vm --all-targets --features executable -- -D warnings` both exit 0.

## Checklist before requesting a review
- [x] Repo forked and branch created from `next` according to naming convention.
- [x] Commit messages and codestyle follow conventions.
- [x] Commits are signed.
- [x] Relevant issues are linked in the PR description.
- [x] Tests added for new functionality.
- [x] Documentation/comments updated according to changes.
- [x] Updated `CHANGELOG.md`

### CODE REVIEW

File: miden-vm/src/cli/prove.rs [hunk 1]
```diff
@@ -104,6 +104,28 @@ impl ProveCmd {
             return Err(Report::msg("The provided file must have a .masm or .masp extension"));
         }
 
+        // The outputs are written after the proof, so if both name the same file the outputs
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: nit: shrink: Please reduce this 11-line comment to the invariant the code cannot show:

`// Prevent outputs from overwriting the proof; compare ASCII case-insensitively for common case-insensitive filesystems.`

The rest paraphrases the next three lines or gives more detail about unsupported aliases than this local guard needs.

File: miden-vm/src/cli/prove.rs [hunk 2]
```diff
@@ -198,4 +220,101 @@ impl ProveCmd {
 
         Ok(())
     }
+
+    /// Returns the path the proof is written to: `--proof` when given, otherwise the program path
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: nit: shrink: Please replace these helper docs with one line each:

`/// Resolves the proof path the same way as ProofFile::write.`

`/// Derives verify's default outputs path from the resolved proof path.`

The function names and bodies already explain the mechanics.

File: miden-vm/src/cli/prove.rs [hunk 3]
```diff
@@ -198,4 +220,101 @@ impl ProveCmd {
...
+            num_outputs: 16,
+            output_file: None,
+            proof_file: proof_file.map(PathBuf::from),
+            hasher: "blake3-256".to_string(),
+            security: "96bits".to_string(),
+            kernel_file: None,
+        }
+    }
+
+    /// With no `--proof`, the proof defaults to the program path, so the outputs land next to the
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: nit: delete: Please remove the narrative doc comments from these five tests. Their names and assertions already state each case, so the comments only repeat the code.

File: miden-vm/tests/integration/cli/cli_test.rs [hunk 4]
```diff
@@ -78,6 +78,129 @@ fn prove_rejects_missing_inferred_inputs_file() {
         .stderr(predicate::str::contains("No such file or directory"));
 }
 
+/// `verify` locates the outputs file by replacing the extension of the proof path, so `prove` has
```
Comment (huitseeker) [inline miden-vm/tests/integration/cli/cli_test.rs]: nit: delete: Please remove the narrative doc comments from these five tests. The test names and assertions already describe the behavior. Keep only comments that explain a non-obvious constraint, such as why the diagnostic must be matched by an unsplittable word.

**Review by huitseeker — COMMENTED** [review]
Thanks! The changes look good but the comments have to be more concise and relevant. They should not be a narration that repeats the code.

**Comment by rayasa07** [comment]
Done, thanks. Kept the unsplittable-word comment since you called that one out as worth keeping. Rebased on next while I was in there, no code changes.

File: miden-vm/src/cli/prove.rs [hunk 5]
```diff
@@ -104,6 +104,19 @@ impl ProveCmd {
             return Err(Report::msg("The provided file must have a .masm or .masp extension"));
         }
 
+        // Prevent outputs from overwriting the proof; compare ASCII case-insensitively for
+        // common case-insensitive filesystems.
+        let proof_path = self.resolved_proof_path();
+        let output_path = self.output_file.clone().unwrap_or_else(|| self.default_output_path());
+        if output_path.as_os_str().eq_ignore_ascii_case(proof_path.as_os_str()) {
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: `--proof ./custom.proof --output custom.proof` passes this raw comparison, then `OutputFile::write` truncates the proof. Could you normalize ordinary `.` and `..` aliases before comparing, or limit this guard to the derived output path it can check reliably?

File: miden-vm/src/cli/prove.rs [hunk 6]
```diff
@@ -198,4 +211,83 @@ impl ProveCmd {
 
         Ok(())
     }
+
+    /// Resolves the proof path the same way as ProofFile::write.
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs:218]: nit: `resolved_proof_path` copies the same default rule from `ProofFile::write`, so these paths can drift again. Could `ProofFile` own one resolver that both the write and output derivation use?

File: miden-vm/src/cli/prove.rs (hunk 5 above)
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: This check still misses lexical aliases. For example, `--proof sub/../same.proof --output same.proof` passes, then the outputs replace the proof. I reproduced this: the command exits 0 and `same.proof` contains JSON outputs. Could we normalize `.` and `..` components before comparing, and add this case to the collision tests?

**Review by huitseeker — CHANGES_REQUESTED** [review]
Lexical aliases such as sub/../same.proof bypass the new collision guard, allowing outputs to overwrite the proof.

**Comment by rayasa07** [comment]
Fixed, thanks. The proof/output comparison now normalizes `.` and `..` components, with a regression test for the reproduced alias. I also moved proof-path resolution into `ProofFile` so writing and output derivation share it. Signed commit `2d69c8c4e` is pushed.

File: miden-vm/src/cli/prove.rs [hunk 7]
```diff
@@ -198,4 +214,119 @@ impl ProveCmd {
...
+        ProofFile::resolve_path(&self.proof_file, &self.program_file)
+    }
+
+    /// Derives verify's default outputs path from the resolved proof path.
+    fn default_output_path(&self) -> PathBuf {
+        self.resolved_proof_path().with_extension("outputs")
+    }
+}
+
+fn paths_are_equal(left: &Path, right: &Path) -> bool {
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs:229]: `paths_are_equal` still treats relative and absolute paths as different. I reproduced `--proof same.proof --output "$PWD/same.proof"`: the command exits 0 and replaces the proof with JSON. Could you resolve both paths from the same working directory and add this case to the test?

**Review by huitseeker — CHANGES_REQUESTED** [review]

**Comment by rayasa07** [comment]
Fixed. Both sides now resolve against the current directory before the comparison, so the `$PWD/same.proof` case is rejected. Added it to the tests.

Went slightly past what you asked since canonicalizing gives it for free. The parent is canonicalized, not just normalized lexically, so a symlinked directory resolves too, and a path that already exists resolves fully, which catches a symlinked proof file. Both have tests. Hard links still compare as distinct and the comment says so.

Rebased on next. The changelog entries had ended up in the 0.31.0 section after 0.32 and 0.33 were cut, so I moved them to unreleased.

File: miden-vm/src/cli/prove.rs [hunk 8]
```diff
@@ -198,4 +215,146 @@ impl ProveCmd {
...
+
+/// Resolves `path` against the current directory so both sides compare in one form.
+///
+/// The file itself usually does not exist yet, so only its parent can be canonicalized; that still
+/// resolves symlinked directories. Hard links compare as distinct.
+fn resolve_for_comparison(path: &Path) -> PathBuf {
+    let lexical = normalize_path(&env::current_dir().unwrap_or_default().join(path));
+
+    // A path that already exists resolves fully, which also follows a symlinked file.
+    if let Ok(canonical) = lexical.canonicalize() {
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: `canonicalize()` fails for a dangling `custom.outputs -> custom.proof` symlink. I reproduced `prove ... --proof custom.proof` exiting 0, then the default outputs write followed the link and replaced the new proof with 201-byte JSON. Please reject this alias and add a CLI regression.

File: miden-vm/src/cli/prove.rs [hunk 9]
```diff
@@ -198,4 +215,146 @@ impl ProveCmd {
...
+    let right = resolve_for_comparison(right);
+    left.as_os_str().eq_ignore_ascii_case(right.as_os_str())
+}
+
+/// Resolves `path` against the current directory so both sides compare in one form.
+///
+/// The file itself usually does not exist yet, so only its parent can be canonicalized; that still
+/// resolves symlinked directories. Hard links compare as distinct.
+fn resolve_for_comparison(path: &Path) -> PathBuf {
+    let lexical = normalize_path(&env::current_dir().unwrap_or_default().join(path));
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: Normalizing before canonicalizing changes `..` across symlinks. I reproduced `--proof real/same.proof --output alias/../same.proof` with `alias -> real/sub`; the guard accepted it, and outputs replaced the proof with 201-byte JSON. Please preserve filesystem resolution order and add this case.

**Review by huitseeker — CHANGES_REQUESTED** [review]

**Comment by rayasa07** [comment]
Fixed. The collision guard now resolves dangling final symlinks and preserves filesystem resolution order across symlinked directories and `..` components. Added CLI regressions for both cases.

Rebased on `next` at `246ef6bb0`; signed head is `d91f439fd`. Verified with nightly rustfmt, clippy (`-D warnings`), the full miden-cli suite (296 passed, 1 ignored), and the focused collision tests.

File: miden-vm/src/cli/prove.rs [hunk 10]
```diff
@@ -198,4 +215,155 @@ impl ProveCmd {
...
+    /// Derives verify's default outputs path from the resolved proof path.
+    fn default_output_path(&self) -> PathBuf {
+        self.resolved_proof_path().with_extension("outputs")
+    }
+}
+
+fn paths_are_equal(left: &Path, right: &Path) -> bool {
+    let left = resolve_for_comparison(left);
+    let right = resolve_for_comparison(right);
+    left.as_os_str().eq_ignore_ascii_case(right.as_os_str())
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs:233]: `eq_ignore_ascii_case` misses Unicode normalization aliases on default APFS. I reproduced `--proof é.proof --output é.proof`: it exits 0 and replaces the proof with JSON. Could you recheck after creating the proof, when `canonicalize` can resolve both spellings, and add this CLI case?

File: miden-vm/src/cli/prove.rs [hunk 11]
```diff
@@ -189,7 +206,7 @@ impl ProveCmd {
             let stack = stack_outputs.get_num_elements(self.num_outputs).to_vec();
 
             // write all outputs to default location if none was provided
-            let default_output_path = self.program_file.with_extension("outputs");
+            let default_output_path = self.default_output_path();
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs:209]: nit: `output_path` was already resolved above, but this branch resolves the default again and duplicates `OutputFile::write`. Could we write once to `output_path`, then keep only the optional stack printing in this branch?

File: miden-vm/src/cli/prove.rs [hunk 12]
```diff
@@ -198,4 +215,155 @@ impl ProveCmd {
...
+    let path = if path.is_absolute() {
+        path.to_path_buf()
+    } else {
+        env::current_dir().unwrap_or_default().join(path)
+    };
+
+    resolve_symlinks(path)
+}
+
+fn resolve_symlinks(mut path: PathBuf) -> PathBuf {
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs:250]: nit: this 39-line symlink walker exists because the check runs before the proof is created. Could we move the check after `ProofFile::write` and use `canonicalize` on both final paths? That removes the custom resolver and also catches filesystem Unicode aliases.

File: miden-vm/tests/integration/cli/cli_test.rs [hunk 13]
```diff
@@ -78,6 +78,251 @@ fn prove_rejects_missing_inferred_inputs_file() {
         .stderr(predicate::str::contains("No such file or directory"));
 }
 
+#[test]
+fn prove_writes_outputs_next_to_a_custom_proof_file() {
+    let working_dir = TempDir::new().unwrap();
```
Comment (huitseeker) [inline miden-vm/tests/integration/cli/cli_test.rs:83]: nit: these ten tests repeat the same setup before their alias-specific steps. Could a `prove_command()` fixture build the working directory and command so each test only contains its unique setup and assertion?

**Comment by rayasa07** [comment]
Done. The check runs after `ProofFile::write` now and canonicalizes both final paths, so the symlink walker is gone and the Unicode case falls out of it. Added that CLI case and a `prove_command()` fixture the alias tests share.

One behaviour change to flag: the guard fires only where the filesystem actually aliases the two paths, so `custom.OUTPUTS` against `custom.outputs` is rejected on APFS and accepted on ext4. The two tests for that probe the filesystem under the test directory and assert whichever branch it implements, rather than assuming. Hard links are still invisible and the comment says so.

Rebased on `next` at 95c4cf717, and the changelog entries moved to 0.34.0 since 0.33 was cut.

File: miden-vm/src/cli/prove.rs [hunk 14]
```diff
@@ -180,22 +184,133 @@ impl ProveCmd {
...
+}
+
+/// Returns true when both paths resolve to the same existing file.
+///
+/// This runs after the proof has been written, so the proof side always resolves. An outputs path
+/// that does not resolve names a file that does not exist yet, which therefore cannot be the proof.
+/// Hard links are the one alias this cannot see: they share no resolved path.
+fn resolve_to_same_file(output_path: &Path, proof_path: &Path) -> bool {
+    match (fs::canonicalize(output_path), fs::canonicalize(proof_path)) {
+        (Ok(output_path), Ok(proof_path)) => output_path == proof_path,
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: Hard links bypass `resolve_to_same_file` because their canonical paths differ, so the final write truncates the new proof and exits successfully. Could this compare file identity, for example with `same-file`, and add a hard-link regression test?

**Review by huitseeker — COMMENTED** [review]
Thanks, I think we're pretty close now, just one more edge case.

**Comment by rayasa07** [comment]
Switched the check to `same_file::is_same_file`, so hard links are caught now. Added a unit test (with a copy as the control) and a CLI regression where the default outputs path is a hard link to the proof. It's a separate commit on top, no rebase this time.

File: miden-vm/src/cli/prove.rs [hunk 15]
```diff
@@ -180,22 +183,147 @@ impl ProveCmd {
...
+        self.resolved_proof_path().with_extension("outputs")
+    }
+}
+
+/// Returns true when both paths name the same existing file.
+///
+/// This runs after the proof has been written, so the proof side always exists. An outputs path
+/// that cannot be opened names a file that does not exist yet, which therefore cannot be the proof.
+fn resolve_to_same_file(output_path: &Path, proof_path: &Path) -> bool {
+    same_file::is_same_file(output_path, proof_path).unwrap_or(false)
```
Comment (huitseeker) [inline miden-vm/src/cli/prove.rs]: `is_same_file` can fail on permission errors, but `unwrap_or(false)` treats those as "different." On Unix, a write-only proof plus hard-linked outputs bypasses this guard, and the outputs write replaces the proof. Could we ignore only `NotFound`, propagate other errors, and add this regression?

**Comment by rayasa07** [comment]
Done. Only NotFound counts as a different file now, any other error fails the command with the proof kept. Added the write-only hard link case as a CLI regression, plus a unit test for the error.
